초기환경 설정

voidmatcha opened the pull request:
> resolve #6 

- [x] CRA가 아닌 Webpack, Babel기반의 React 환경 구축
- [x] ESLint 환격 적용
- [x] Storybook  환경 적용
- [x] Jest와 React Testing Library 기반의 유닛테스트 환경 적용
- [x] E2E 테스트를 위한 Cypress 환경 적용
- [x] react-router-dom 설정
- [x] Redux 설정
- [x] GitHub Action 설정
- [x] styled-components, SCSS 설정

Commit:
env: ESLint 설정 적용 (#6)
eslint --init 후 react/jsx-filename-extension rule을 추가했다.

Commit:
env: Storybook 환경 적용 (#6)
https://velog.io/@telung123/Storybook
위 문서를 참고하여 환경을 적용하였다.

Commit:
env: Jest와 React Testing Library 기반의 유닛테스트 환경 적용 (#6)
환경을 적용한 후 각 파일별로 유닛테스트를 작성하였다.

Commit:
env: E2E 테스트를 위한 Cypress 환경 적용 (#6)
https://github.com/adhrinae/react-testing-tutorial-kr/blob/master/translations/ch10.md
위 게시글을 참고하여 환경을 적용하였다.
제대로된 테스트는 아직 작성하지 않았다.
npm run test:cypress 명령어로 실행 가능하다.

Commit:
env: react-router-dom 설정 (#6)
webpack.config.js의 devserver 항목 옵션 중 historyApiFallback: true 를
추가해야 Cannot GET / 에러가 해결된다.
App.js에 react-router-dom 기반의 코드를 작성하였고 유닛테스트를
추가하였다.

Commit:
env: Redux 설정 (#6)
redux-toolkit의 createSlice API를 이용해 ducks pattern으로 세팅하였다.
Counter 예제를 만들어 놓았고 이에대한 테스트코드를 작성하였다.

Commit:
env: GitHub Action 설정 (#6)
master 브랜치에 push가 반영되거나 PR이 merge되면 Test가 출력되는
Workflow를 정의하였다.

Commit:
env: styled-components, SCSS 설정 (#6)
https://www.onlinewebfonts.com/download/67680f2eb947c5cbd58b40961b4a61e9
Apple SD Gothic Neo Bold woff 파일은 위 링크를 통해 구하였다.
styled-components 에서 local woff 파일을 읽지 못하는 문제가 발생하여 SCSS
환경을 구성하고 index.scss 파일로 font-face 를 가져오도록 구성하였다.
index.scss 의 font-face 를 styled-components 에서 사용할 수
있다.(Button.js 확인)
SCSS 와 woff 파일을 읽을 수 있도록 webpack.config.js 를 수정하였다.

SkynI25 commented:
혹시 개인적으로 궁금한건데요, package-lock.json 파일도 gitignore로 무시하지 않고 업로드 해주는게 좋은가요?

SkynI25 commented on src/App.js:
```diff
@@ -1,11 +1,17 @@
-    </div>
+    <>
+      <Switch>
+        {/* <GlobalFonts /> */}
```
해당 부분은 사용하지 않으면 지우는게 어떨까요? ㅎㅎ

voidmatcha replied:
앗 제가 확인을 제대로 안했군요.. 다음 커밋에 지우도록 하겠습니다!

voidmatcha commented:
> 혹시 개인적으로 궁금한건데요, package-lock.json 파일도 gitignore로 무시하지 않고 업로드 해주는게 좋은가요?

네네 의존성이 보장되게끔 해주는 역할을 합니다. 아래 링크에 자세히 설명되어 있어요!
https://hyunjun19.github.io/2018/03/23/package-lock-why-need/

SkynI25 commented on public/fonts/fonts.js:
```diff
@@ -0,0 +1,15 @@
+    url(${AppleSDGothicNeoBoldWoff}) format('woff');
+    font-weight: 300;
+    font-style: normal;
+  }
```
fonts.js 파일은 어느시점에서 사용되는 건지 설명해주실 수 있을까요?

voidmatcha replied:
아 이건 제가 SCSS없이 styled-components 만으로 font-face를 지정하는 경우를 확인해보기위해 만들어둔 파일이었는데 지우질 못했네요. 혹시 모르니 다시 시도해보고 안될경우 다음 커밋에 지워놓도록 하겠습니다!

SkynI25 reviewed (commented):
전체적으로 확인했습니다 :) 환경설정 작업 수행하시느라 고생 많으셨습니다. 전반적으로 잘못되었다고 느낀 부분은 없었고 storybook이나 cypress 테스트 도구과 같이 새로이 접하는 부분들이 있어 개인적으로 더 파보고 싶어졌습니다 ㅎㅎ

voidmatcha commented on package.json:
```diff
@@ -25,14 +29,48 @@
+    "redux": "^4.0.5",
+    "redux-thunk": "^2.3.0",
+    "sass-loader": "^10.0.2",
+    "style-loader": "^1.3.0",
```
다시 확인해보니 Webpack 설정에 필요한 모듈들(css-loader, node-sass, sass-loader, style-loader)이 dependencies에 적용되어있었네요. 다음 커밋에 devDependencies로 옮겨놓겠습니다.

Lavegaa commented on src/utils.js:
```diff
@@ -0,0 +1,4 @@
+/* eslint-disable import/prefer-default-export */
+export function get(key) {
+  return (obj) => obj[key];
+}
```
utils에는 앞으로 어떤 함수들이 추가될 예정인가요?

Lavegaa replied:
유틸은 utils폴더를 하나 만들어야할것같아요
reducers와 store도 store폴더를 만드는것이 좋을 것 같습니다.

voidmatcha replied:
utils에는 구현하면서 필요한 스니펫들이 들어갈 예저입니다. 말씀하신대로 폴더로 따로 빼도록 하겠습니다. 다음 커밋에 반여하겠습니다.

SkynI25 replied:
`스니펫 : 스니펫은 재사용 가능한 소스 코드, 기계어, 텍스트의 작은 부분을 일컫는 프로그래밍 용어다.`
ㅎㅎ 해당 용어들도 많이 참고하겠습니다..두분께서 많이 알고 계시네요!

Lavegaa commented on src/reducer.js:
```diff
@@ -0,0 +1,3 @@
+import reducer from './slice';
+
+export default reducer;
```
이 코드는 왜 있는건가요??
이름을 reducer로 맞춰주기 위해서인가요? 아니면 테스트를 위한건가요??

voidmatcha replied:
사실 필요가 없습니다. 지우도록 하겠습니다.

Lavegaa commented on src/slice.js:
```diff
@@ -0,0 +1,24 @@
+  dispatch(setNumber({ number: 0 }));
+};
+
+export default reducer;
```
redux-tools라이브러리를 처음봐서 이해하는데 시간이 좀 걸렸네요.
해당 라이브러리 관련 스터디 한번 있으면 좋을 것 같아요.

- 공식 페이지 https://redux-toolkit.js.org/
- 소개글 https://jbee.io/react/introduce-redux-starter-kit/#redux-actions-%ED%8F%AC%EC%9A%A9

voidmatcha replied:
네네 한번 시간맞춰서 한번 진행해보도록 해요.

Lavegaa commented on src/slice.js:
```diff
@@ -0,0 +1,24 @@
+import { createSlice } from '@reduxjs/toolkit';
+
+const { actions, reducer } = createSlice({
```
네이밍은 예시로 actions랑 reducer로 하신건가요?

sjh2428 replied:
네이밍을 **예시로** `actions`, `reducer`라고 하신 것이 아니라 **사용되어온 이름**인 것으로 이해하시면 좋을 것 같습니다!

리덕스에서는 하나의 store에서 상태를 관리하는데 store에 저장되어있는 상태에 대해
무언가를 하고 싶다(예를 들어 숫자를 +1 또는 -1 하고 싶다) 할 때 어떠한 트리거를 발생시켜야 하는데
이러한 트리거를 `action`이라 명명하였으며 트리거를 당기는 행위를 "`action`을 발행한다" 라고 할 수 있습니다.
다시 거꾸로 말해 보자면 "action을 발행한다"는 것은 store에 있는 state 값을 변경해달라는 일종의 트리거라고 생각하시면 될 것 같습니다.

위에서 설명드린 action은 다음과 같은 데이터를 가지고 발행합니다.
```js
{
    type: "액션의 종류를 한번에 식별할 수 있는 문자열 혹은 심볼",
    payload: "액션의 실행에 필요한 임의의 데이터",
}
```

보통 이렇게 사용합니다.
```js
export const ADD_VALUE = "ADD_VALUE";
export const addValue = (payload) => ({ type: ADD_VALUE, payload });
```

redux store에는 이전 state값이 저장되어 있는데, 이전 state값과 action으로 넘어온 데이터(type, payload)를 가지고 store에 있는 state 값을 새로 쓰도록 하는 로직을 담고있는 녀석의 이름이 `reducer`입니다.
toolkit을 사용하지 않은 코드는 다음과 같습니다.

```js
import { ADD_VALUE } from './actions';

const initialState = {
    value: 0
};
​
export default (state = initialState, action) => {
    const { type, payload } = action;

    switch (type) {
        case ADD_VALUE:
            return { ...state, value: state.value + payload };
        default:
            return state;
    }
}
```

저는 개인적인 생각으로는 redux를 사용할 때는 typescript를 사용한다면 좋을 것 같습니다..!
-> (물론 약 2달이 남은 현 시점에서 typescript와 redux, redux-saga를 취준과 병행해서 하라는 말은 무리이긴 합니다..)

위의 코드처럼 `ADD_VALUE`라는 이름의 액션이 들어오면 value 값을 `value: state.value + payload` 갱신하게 되는데,
payload에 string 값이 들어가게 된다면 실행을 해본 후에야 에러를 발견하게 됩니다. 개발자가 신경쓰면서 잘 코딩을 한다면
문제가 없지만 typescript를 쓴다면 컴파일 단계에서 에러를 잡아주기 때문에 위의 문제를 코딩할 때 발견할 수 있습니다.
이에 더하여 js에서 코드들이 모듈화가 되다 보면 제대로 사용할 수 없었던 **자동완성 기능**이 사용 가능합니다!🎉✨🎊

또한 redux를 많이 사용해보시지 않으셨다면(많이 사용해 보셨다면 넘기셔도 됩니다!) toolkit을 이용하지 않고
먼저 redux를 날것(?)으로 구현해 보시고 추후에 toolkit과 같은 라이브러리의 도움을 받으셨으면 좋겠습니다.

백엔드 개발자가 많은 참견과 지적일 수 있겠지만 모쪼록 도움이 되셨으면 하는 마음입니다..!

아시는 것을 짚은 것은 아닌지 조심스럽네요 ㅎㅎ..

찾아보니 Redux의 전체적인 흐름을 잘 설명해 놓은 블로그와
저도 많은 도움을 받았던 벨로퍼트의 리덕스(+리덕스 사가)가 보여 링크 놓고 가겠습니다 :D

[Redux의 전체적인 흐름 Reference](https://medium.com/@ca3rot/%EC%95%84%EB%A7%88-%EC%9D%B4%EA%B2%8C-%EC%A0%9C%EC%9D%BC-%EC%9D%B4%ED%95%B4%ED%95%98%EA%B8%B0-%EC%89%AC%EC%9A%B8%EA%B1%B8%EC%9A%94-react-redux-%ED%94%8C%EB%A1%9C%EC%9A%B0%EC%9D%98-%EC%9D%B4%ED%95%B4-1585e911a0a6)

[벨로퍼트의 리덕스](https://react.vlpt.us/redux/01-keywords.html)

Lavegaa replied:
음 제가 말씀드린건 slice.js안에서 여러 reducer를 작성할 때를 말씀드린거에요. 
아래 코드는 redux-tools 튜토리얼에 있는 코드입니다.
https://redux-toolkit.js.org/api/createSlice
```js
import { createSlice, createAction } from '@reduxjs/toolkit'
import { createStore, combineReducers } from 'redux'

const incrementBy = createAction('incrementBy')
const decrementBy = createAction('decrementBy')

const counter = createSlice({
  name: 'counter',
  initialState: 0,
  reducers: {
    increment: (state) => state + 1,
    decrement: (state) => state - 1,
    multiply: {
      reducer: (state, action) => state * action.payload,
      prepare: (value) => ({ payload: value || 2 }), // fallback if the payload is a falsy value
    },
  },
  // "builder callback API", recommended for TypeScript users
  extraReducers: (builder) => {
    builder.addCase(incrementBy, (state, action) => {
      return state + action.payload
    })
    builder.addCase(decrementBy, (state, action) => {
      return state - action.payload
    })
  },
})

const user = createSlice({
  name: 'user',
  initialState: { name: '', age: 20 },
  reducers: {
    setUserName: (state, action) => {
      state.name = action.payload // mutate the state all you want with immer
    },
  },
  // "map object API"
  extraReducers: {
    [counter.actions.increment]: (
      state,
      action /* action will be inferred as "any", as the map notation does not contain type information */
    ) => {
      state.age += 1
    },
  },
})
```
보통은 counter나 user처럼 기능에 맞게 네이밍을 만들어 counter.actions나 counter.reducer로 사용하는데, actions나 reducer로 만드신게 예시인지 따로 패턴이 있는지 여쭤본거에요. 아니면 기능에 따라 여러 파일로 나눠서 작성해 하나의 파일에 하나의 reducer만 있는건가요?

sjh2428 replied:
아하..! 제가 잘못 짚었나봅니다!

제 생각에는 찬호님 말씀처럼 기능에 따라 여러 파일로 나눠서 작성해 하나의 파일에 하나의 reducer만 있는 것이 좋을 것 같습니다!

voidmatcha replied:
네네 찬호님 말씀대로 진행하는게 좋을 것 같습니다. 반영해 놓겠습니다.

Lavegaa replied:
다른 패턴이 있나 궁금해서 여쭤본거였어요!
저희는 하나의 파일에 하나의 리듀서로 가는걸로 알고있겠습니다!

voidmatcha replied:
아 그리고 제가 네이밍을 했다기보다는 createSlice에서 반환되는 객체의 값 중 action과 reducer를 구조분해 할당해서 가져온 것에 가깝습니다!

SkynI25 replied:
재호님이 redux에 대해서 아주 굴직하고 자세히 설명해주셨네요 :) 감사합니다. 저도 첫 리뷰할 때는 간과했는데 redux-toolkit 에 대해서 제대로 알아야 되겠다는 생각이 들었습니다.

Lavegaa commented on src/components/Button.test.js:
```diff
@@ -0,0 +1,61 @@
+      expect(dispatch).toBeCalledTimes(1);
+    });
+  });
+});
```
테스트 코드도 작성해본적이 없어서 이것도 스터디 한번 가능할까요?

voidmatcha replied:
넵 이부분도 시간 맞춰서 진행해보면 좋을 듯 합니다.

Lavegaa committed:
env: api연동 설정 (#6)

Lavegaa commented:
api연동하는 부분 커밋했습니다.
axios 라이브러리 기반으로 작성했습니다.

Commit:
refactor(env): PR의 코멘트 내용을 바탕으로 리팩토링 진행 (#7)
- App.js 의 필요없는 주석과 fonts.js 파일을 삭제하였다.
- devDependencies에 있어야 할 모듈들을 원위치 시켰다.
- utils, store 폴더를 만들어 파일 구조들을 정리하였다.

SkynI25 reviewed (approved):
확인했습니다 👍
